File: windows-native/PitlaneDanmaku.Windows/Services/LogService.cs
using System.Collections.ObjectModel;

namespace PitlaneDanmaku.Windows.Services;

public sealed class LogService
{
    private readonly object _gate = new();
    private readonly List<string> _entries = [];

    public event Action<string>? EntryAdded;

    public IReadOnlyList<string> Snapshot
    {
        get
        {
            lock (_gate)
            {
                return new ReadOnlyCollection<string>(_entries.ToList());
            }
        }
    }

    public void Info(string message) => Add("INFO", message);

    public void Warn(string message) => Add("WARN", message);

    public void Error(string message) => Add("ERROR", message);

    public void Add(string level, string message)
    {
        var entry = $"[{DateTimeOffset.Now:HH:mm:ss}] {level,-5} {message}";
        lock (_gate)
        {
            _entries.Add(entry);
            if (_entries.Count > 1500)
            {
                _entries.RemoveRange(0, 300);
            }
        }

        EntryAdded?.Invoke(entry);
    }

    public void Clear()
    {
        lock (_gate)
        {
            _entries.Clear();
        }
    }
}

File: macos-native/Sources/PitlaneDanmakuMac/LogService.swift
import Foundation

final class LogService {
    private let queue = DispatchQueue(label: "PitlaneDanmaku.LogService")
    private var entries: [String] = []

    var onEntry: ((String) -> Void)?

    var snapshot: [String] {
        queue.sync { entries }
    }

    func info(_ message: String) {
        add(level: "INFO", message)
    }

    func warn(_ message: String) {
        add(level: "WARN", message)
    }

    func error(_ message: String) {
        add(level: "ERROR", message)
    }

    func clear() {
        queue.sync {
            entries.removeAll()
        }
    }

    private func add(level: String, _ message: String) {
        let formatter = DateFormatter()
        formatter.dateFormat = "HH:mm:ss"
        let entry = "[\(formatter.string(from: Date()))] \(level.padding(toLength: 5, withPad: " ", startingAt: 0)) \(message)"

        queue.sync {
            entries.append(entry)
            if entries.count > 1500 {
                entries.removeFirst(min(300, entries.count))
            }
        }

        onEntry?(entry)
    }
}
